src/lib/ares_dns_write.c

```diff
@@ -1052,3 +1053,28 @@ void ares_dns_record_write_ttl_decrement(ares_dns_record_t *dnsrec,
   }
   dnsrec->ttl_decrement = ttl_decrement;
 }
+
+/* Write a DNS record representing a query for a single name, but temporarily
+ * overwrite the name with an alternative name before doing so. This is used
+ * as a helper function in ares_search().
+ */
+ares_status_t ares_dns_write_query_altname(ares_dns_record_t *dnsrec,
+                                           char *altname, unsigned char **buf,
+                                           size_t *buflen)
+{
+  char         *qname;
+  ares_status_t status;
+
+  if (ares_dns_record_query_cnt(dnsrec) != 1) {
+    return ARES_EBADQUERY;
+  }
+
+  qname = dnsrec->qd[0].name;
+  if (altname != NULL) {
+    dnsrec->qd[0].name = altname;
+  }
+  status = ares_dns_write(dnsrec, buf, buflen);
+  dnsrec->qd[0].name = qname;
+
+  return status;
+}
```

Comment:
I guess this is more of a question than anything, but is there any concern about this being used for "responses" where we have RRs and might need to replace the name in the RR?  I'm not sure if there's a usecase for this, but we might want to at least document that it only replaces the name in the query at this time and doesn't do anything for RRs.

Reply:
I'm not concerned about this function being used for responses, since it is private and only used by `ares_search_dnsrec()`. The name also contains `query` so I think it's unlikely to be misused.

Potentially we should get it to look at extra RRs on queries too. For example, in the EDNS0 OPT RR case we will have RRs on queries in the additional section (but they will all use the root domain `""` per RFC-6891). However, it's not clear to me whether this is necessary or correct in all usecases.

Also there could be complications around subdomains:
- Suppose we are rewriting `example.com` to `example.com.searchdomain.org`.
- Should `test.example.com` in an RR be re-written to `test.example.com.searchdomain.org`?
- Should `example.com.otherdomain.gov` in an RR be re-rewritten to `example.com.searchdomain.org.otherdomain.gov`?
- Is there even a usecase for this?

In any case, I think clearly documenting how the search domains affect the query (+RRs) in `ares_search_dnsrec()` is important, and I will push a markup to clarify this.

How about we make `ares_search_dnsrec()` only change the name in the query, and not affect any RRs? More complicated usage could be handled in user-code by calling `ares_send()` directly (which users must be using currently).